Increase NV2A_MAX_BATCH_LENGTH to 131071

apexad opened the pull request:
This should be good to merge. I added a very basic unit test above line 5531 in nv2a.c:

if (pg->inline_elements_length > 65536) { printf("value is %d",pg->inline_elements_length); }

Looks like the value gets to around 80,070 in Aeon Flux.  In testing other games as well, I have seen no regression with increasing this value.

This quick patch should allow other games that only reach the intro to go in game.

apexad committed:
Increase NV2A_MAX_BATCH_LENGTH to 131071 from 65535

espes commented:
What does the draw block look like when batches are this big?

JayFoxRox commented:
We should not use hacks like this. Figure out why it has so many vertices. The current limits should be actual hardware limits [and whats used by the D3D driver]. Try dumping the vertex buffers and convince me that this is how to do it properly. [or test this on real hw - it might throw an interrupt or something?]

apexad commented:
Just a note, I gave the commit a cleaner message (removed the Aeon Flux Fix part).
I've run apitrace and here is the warnings in the replay:

`WARNING: Could not find vertex shader attribute 'v6' to match BindAttributeLocation request.`
(and other lines similar that with different v# values)
and also items like this:
`720898 @1 glFrameTerminatorGREMEDY()`
`720898: warning: could not infer drawable size (glViewport never called)`

I'm not sure how to look at the 'draw block' or how to 'dump the vertex buffers' or how to check this on real hardware.  I'm highly willing to learn if either of you can point me in the right direction or help me to understand the above errors so I can make corrections to this pull request.

I will continue to research this further and try to understand the messages given by apitrace.

espes commented:
Enable the DEBUG_NV2A define and look at the output, particularly the commands inside the draw begin/end block.

apexad commented:
Perhaps these may help explain why this change fixes Aeon Flux.
Actual Xbox:
<img width="200" alt="screen shot 2015-09-13 at 9 57 33 pm" src="https://cloud.githubusercontent.com/assets/1437332/9842505/afed1954-5a63-11e5-956b-264ef40b2826.png">
xqemu:
<img width="200" alt="screen shot 2015-09-13 at 9 52 17 pm" src="https://cloud.githubusercontent.com/assets/1437332/9842506/b3f13594-5a63-11e5-972d-408b8ec92242.png">

I believe this is the draw block?
nv2a: dma_map 3d, 0, 0 7ffafff
nv2a: dma_map 3d, 0, 0 7ffafff
nv2a: pgraph_set_surface_dirty(1, 1) -- 1 1
MAX Check: value is 80066
MAX Check: value is 80068
MAX Check: value is 80070
nv2a: pgraph method (0) 0x1800 \* 40035
nv2a: pgraph method (0): 0x97 -> 0x1808 (0x6a9c)
nv2a: pgraph method (0): 0x97 -> 0x17fc (0x0)

Max Check is my printf currently in if (pg->inline_elements_length > 80064)
I do find it interesting that pg->inline_elements_length is going up by 2 each time

espes commented:
What are all the methods between the two 0x17fc?

apexad commented:
nv2a: pgraph method (0): 0x97 -> 0x17fc (0x0)
set begin end
nv2a: dma_map 3d, 0, 0 7ffafff
nv2a: dma_map 3d, 0, 0 7ffafff
nv2a: converted 1
nv2a: dma_map 3d, 0, 0 7ffafff
nv2a: converted 2
nv2a: dma_map 3d, 0, 0 7ffafff
nv2a: converted 3
nv2a: dma_map 3d, 0, 0 7ffafff
nv2a: converted 4
nv2a: dma_map 3d, 0, 0 7ffafff
nv2a: dma_map 3d, 0, 0 7ffafff
nv2a: dma_map 3d, 0, 0 7ffafff
nv2a: pgraph_set_surface_dirty(1, 1) -- 1 1
nv2a: pgraph method (5): 0x44 -> 0x0310 (0x7c2dabe4)
nv2a:     unhandled  (0x44 0x00000310)
nv2a: pgraph method (0): 0x97 -> 0x1d70 (0x13b9)
nv2a: dma_map 3d, 0, 0 7ffafff
nv2a: read_surface color 0x0 - 0x7ffafff, (0x3cb0000 - 0x12c000, 0 0, 640 480, 2560) - 0 0 0 ff
nv2a: dma_map 3d, 0, 0 7ffafff
nv2a: read_surface zeta 0x0 - 0x7ffafff, (0x3a58000 - 0x12c000, 0 0, 640 480, 2560) - 0 ff ff ff
nv2a: dma_map 3d, 2, 3fdd000 20
nv2a: pgraph method (0): 0x97 -> 0x1d90 (0x0)
nv2a: pgraph method (0): 0x97 -> 0x1d90 (0x0)
nv2a: pgraph method (0): 0x97 -> 0x1720 (0x2e48e14)
nv2a: pgraph method (0): 0x97 -> 0x1724 (0x2e48e2c)
nv2a: pgraph method (0): 0x97 -> 0x1728 (0x2e48e30)
nv2a: pgraph method (0): 0x97 -> 0x172c (0x2e48e28)
nv2a: pgraph method (0): 0x97 -> 0x1730 (0x2e48e34)
nv2a: pgraph method (0): 0x97 -> 0x1734 (0x2e48e38)
nv2a: pgraph method (0): 0x97 -> 0x173c (0x2e48e20)
nv2a: pgraph method (0): 0x97 -> 0x1740 (0x2e48e24)
nv2a: pgraph method (0): 0x97 -> 0x17fc (0x5)
set begin end
nv2a: dma_map 3d, 0, 0 7ffafff
nv2a: dma_map 3d, 0, 0 7ffafff
nv2a: pgraph_set_surface_dirty(1, 1) -- 1 1
MAX Check: value is 80066
MAX Check: value is 80068
MAX Check: value is 80070
nv2a: pgraph method (0) 0x1800 * 40035
nv2a: pgraph method (0): 0x97 -> 0x1808 (0x6a9c)
nv2a: pgraph method (0): 0x97 -> 0x17fc (0x0)
set begin end

apexad commented:
nv2a: pgraph method (0): 0x97 -> 0x17fc (0x0)
Set Begin End
nv2a: dma_map 3d, 0, 0 7ffafff
nv2a: dma_map 3d, 0, 0 7ffafff
nv2a: converted 1
nv2a: dma_map 3d, 0, 0 7ffafff
nv2a: converted 2
nv2a: dma_map 3d, 0, 0 7ffafff
nv2a: converted 3
nv2a: dma_map 3d, 0, 0 7ffafff
nv2a: converted 4
nv2a: dma_map 3d, 0, 0 7ffafff
nv2a: dma_map 3d, 0, 0 7ffafff
nv2a: dma_map 3d, 0, 0 7ffafff
nv2a: pgraph_set_surface_dirty(1, 1) -- 1 1
nv2a: pgraph method (5): 0x44 -> 0x0310 (0x7c2e57dc)
nv2a:     unhandled  (0x44 0x00000310)
nv2a: pgraph method (0): 0x97 -> 0x1d70 (0x13d7)
nv2a: dma_map 3d, 0, 0 7ffafff
nv2a: read_surface color 0x0 - 0x7ffafff, (0x3b84000 - 0x12c000, 0 0, 640 480, 2560) - 0 0 0 ff
nv2a: dma_map 3d, 0, 0 7ffafff
nv2a: read_surface zeta 0x0 - 0x7ffafff, (0x3a58000 - 0x12c000, 0 0, 640 480, 2560) - 0 ff ff ff
nv2a: dma_map 3d, 2, 3fdd000 20
nv2a: pgraph method (0): 0x97 -> 0x1d90 (0x0)
nv2a: pgraph method (0): 0x97 -> 0x1d90 (0x0)
nv2a: pgraph method (0): 0x97 -> 0x1720 (0x2e48e14)
nv2a: pgraph method (0): 0x97 -> 0x1724 (0x2e48e2c)
nv2a: pgraph method (0): 0x97 -> 0x1728 (0x2e48e30)
nv2a: pgraph method (0): 0x97 -> 0x172c (0x2e48e28)
nv2a: pgraph method (0): 0x97 -> 0x1730 (0x2e48e34)
nv2a: pgraph method (0): 0x97 -> 0x1734 (0x2e48e38)
nv2a: pgraph method (0): 0x97 -> 0x173c (0x2e48e20)
nv2a: pgraph method (0): 0x97 -> 0x1740 (0x2e48e24)
nv2a: pgraph method (0): 0x97 -> 0x17fc (0x5)
Set Begin End
nv2a: dma_map 3d, 0, 0 7ffafff
nv2a: dma_map 3d, 0, 0 7ffafff
nv2a: pgraph_set_surface_dirty(1, 1) -- 1 1
MAX Check: value is 80066
MAX Check: value is 80068
MAX Check: value is 80070
nv2a: pgraph method (0) 0x1800 \* 40035
nv2a: pgraph method (0): 0x97 -> 0x1808 (0x6a9c)
nv2a: pgraph method (0): 0x97 -> 0x17fc (0x0)
Set Begin End

JayFoxRox commented:
Just checked.. number of indices for indexed drawing [inline_elements = 0x1800 + 0x1808] can be a max count of 0x2000000.
~~The other limit must be for another draw type [probably inline_array, maybe inline_buffer].~~
I believe there is no limit for the other methods. The limit we are currently using is the amount of vertices per FIFO method. Only DRAW_ARRAYS is limited to a "START_INDEX" with 16 bit address, Also "START_INDEX + COUNT **\+ 1**" seems to be limited to 16 bit values [despite being uploaded seperately as "COUNT" (which seems to be "actual count **\- 1**", so we might be missing vertices at the moment?)!]
As methods can be called repeatedly it's possible to upload more vertices. So we probably have to realloc buffers.

apexad commented:
0x2000000 as in 33554432 decimal?  Should I update NV2A_MAX_BATCH_LENGTH to 0x2000000 for this pull request?

espes commented:
0x2000000 would be 64 MiB of (16-bit) indicies...

We probably need to refactor all drawing to not buffer everything.

JayFoxRox commented:
Yeah, we need dynamicly allocated buffers. We shouldn't have any limits on our draw calls [except assertions on the DRAW_ARRAYS parameters to confirm if the D3D check is correct + we should test that one on hardware].
**This PR shouldn't be a simple constant change anymore.**
Wether we enforce the limit for FIFO pushes is a different story, but those might only be D3D limits.

Low priority, only affects:
- Aeon Flux
- Jade Empire

[we can possibly fix a handful of other games by refactoring the drawing code: empty draw calls & different draw calls per begin/end]

espes commented:
I'm actually fine doubling it now. While we're using an arbitrary limit it might as well be a limit that fixes this game

JayFoxRox commented:
Sounds good to me, we should still keep this discussion in mind for future changes.
